feat(marketing-analytics): columns from the FE to BE and more

jabahamondes opened the pull request:
## Changes

- Changed MarketingAnalyticsOrderBy type from number to string for better compatibility with the DataTable.
- Introduced ColumnFeature enum to manage sortable and editable columns in DataTable.
- Updated Query and DataTable components to support custom column features.
- Refactored logic to handle dynamic conversion goals and orderBy in MarketingAnalyticsTable
- Allowing to hide columns
- Introducing the config columns modal
- Fixing small bug when clicking reloading button should reload the marketing cost chart
- Column configuration persistance in url
- Allowing to pin columns in marketing table
- Fix padding in row over the table for the reload and export buttons from the table


## How did you test this code?
Manually and added/updated unit tests


Now the marketing table:
(temporary removed the new option from the dropdown from https://github.com/PostHog/posthog/pull/35387 to save the conversion goal)
<img width="1215" height="378" alt="image" src="https://github.com/user-attachments/assets/12ff38d6-4ebc-4078-a831-da6c212b9b58" />

configure column modal:
<img width="636" height="732" alt="image" src="https://github.com/user-attachments/assets/a1b2e865-d094-4688-92aa-b385eaac28f3" />


Following the activity tab table style but with less options:
<img width="1264" height="627" alt="image" src="https://github.com/user-attachments/assets/2778fdab-3a34-4f18-baaa-064facc826ad" />

https://github.com/user-attachments/assets/4506e4e2-4ccb-4e75-9e22-6f7bc12ada1a

Commit:
Squashed commit of the following:

commit d51fd370d3942d791a177e3cbe784f768df2bfd0
Author: Javier Bahamondes <javierbahamondes@Javiers-MacBook-Pro.local>
Date:   Fri Jul 25 16:17:07 2025 -0400

    reverting change and many more

jabahamondes commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/logic/utils.test.ts:
```diff
@@ -0,0 +1,231 @@
+            expect(result).toEqual(['b', 'c', 'a'])
+        })
+
+        it('should preserve order within preference groups', () => {
+            const array = ['a', 'b', 'c', 'd']
+            const preference = ['c', 'a']
+
+            const result = orderArrayByPreference(array, preference)
+
+            expect(result).toEqual(['a', 'c', 'b', 'd'])
+        })
```
it's fine

jabahamondes commented on frontend/src/lib/lemon-ui/LemonTable/LemonTable.tsx:
```diff
@@ -156,7 +146,7 @@ export function LemonTable<T extends Record<string, any>>({
             }
         },
-        [location, searchParams, hashParams, push]
+        [location, searchParams, hashParams, push, useURLForSorting, onSort, currentSortingParam]
```
I going to revert this actually

jabahamondes commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/components/MarketingAnalyticsTable/MarketingAnalyticsColumnConfigModal.tsx:
```diff
@@ -0,0 +1,385 @@
+    const { query: rawQuery, sortedColumns } = useValues(marketingAnalyticsTableLogic)
+    const { setQuery } = useActions(marketingAnalyticsTableLogic)
+    // useRef to avoid re-rendering the component when we pin columns but after we rerender the component we need to keep the sorted columns
+    const staticSortedColumns = useRef(sortedColumns)
```
It's on purpose I don't want the UI to change while the user clicks

jabahamondes commented on frontend/src/lib/lemon-ui/LemonTable/columnUtils.tsx:
```diff
@@ -70,3 +70,96 @@ export function statusColumn<T extends { enabled: boolean }>(): LemonTableColumn
+    let leftPosition = 0
+    if (isPinned && pinnedColumnWidths && allColumns) {
+        // Find all pinned columns that come before this one in the table
+        const pinnedColumnsBeforeThis = pinnedColumns.filter((pinnedKey) => {
+            const pinnedKeyIndex = allColumns.findIndex((col) => (col.key || col.dataIndex) === pinnedKey)
+            const thisColumnIndex = allColumns.findIndex((col) => (col.key || col.dataIndex) === columnKey)
+            return pinnedKeyIndex < thisColumnIndex
+        })
```
Refactored

jabahamondes commented on frontend/src/lib/lemon-ui/LemonTable/columnUtils.tsx:
```diff
@@ -70,3 +70,96 @@ export function statusColumn<T extends { enabled: boolean }>(): LemonTableColumn
+        })
+
+        // Sum up widths of pinned columns that come before this one
+        for (const beforeKey of pinnedColumnsBeforeThis) {
+            const beforePinnedIndex = pinnedColumns.indexOf(beforeKey)
+            if (beforePinnedIndex >= 0) {
+                leftPosition += pinnedColumnWidths[beforePinnedIndex] || DEFAULT_COLUMN_WIDTH
+            }
+        }
```
Refactored

jabahamondes commented on frontend/src/queries/nodes/DataTable/DataTable.tsx:
```diff
@@ -222,57 +240,74 @@ export function DataTable({
+                                            const source = query.source as EventsQuery
+                                            const columns = columnsInLemonTable ?? getDataNodeDefaultColumns(source)
+                                            const isAggregation = isHogQLAggregation(hogQl)
+                                            const isOrderBy = source.orderBy?.[0] === key
+                                            const isDescOrderBy = source.orderBy?.[0] === `${key} DESC`
```
This is unchanged just wrapped in a condition

Commit:
Prs comments

Commit:
Merge remote-tracking branch 'origin/master' into refactor/using-select-columns-from-FE-to-BE-squashed

jabahamondes commented on frontend/src/lib/lemon-ui/LemonTable/columnUtils.tsx:
```diff
@@ -70,3 +70,100 @@ export function statusColumn<T extends { enabled: boolean }>(): LemonTableColumn
+}
+
+/**
+ * Determine the column's key, using `dataIndex` as fallback.
```
This was just moved

jabahamondes commented on frontend/src/queries/nodes/DataTable/DataTable.tsx:
```diff
@@ -222,57 +241,74 @@ export function DataTable({
-                        <LemonDivider />
-                        {canSort && key !== 'person.$delete' && key !== 'person' ? (
+                        {columnFeatures.includes(ColumnFeature.canEdit) && (
+                            <>
```
The main change here is that I wrapped into the condition with is by default enabled

jabahamondes replied:
I recommend to hide whitespaces here to see better the diff

jabahamondes commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/components/MarketingAnalyticsTable/MarketingAnalyticsTable.tsx:
```diff
@@ -1,219 +1,58 @@
-const QUERY_ORDER_BY_START_INDEX = 1
-
 export const MarketingAnalyticsTable = ({ query, insightProps }: MarketingAnalyticsTableProps): JSX.Element => {
-    const { setMarketingAnalyticsOrderBy, clearMarketingAnalyticsOrderBy, saveDraftConversionGoal } =
```
All of this can be cleaned because now we use directly the query with the table logic

Commit:
feat: add support for custom column features in DataTable and Query components

- Introduced  prop in  to allow passing custom column features to the DataTable.
- Updated  to support existing column features by default if not provided.
- Adjusted  to utilize the new  prop.

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

lricoy commented:
@jabahamondes generally, LGTM, could you add a storybook update to showcase the pinned bits? I will try running it locally because I don't have a lot of context on the DataTable changes. 

You will probably want a review from @PostHog/team-product-analytics on this one

Commit:
fixing select case when empty

Commit:
Merge branch 'refactor/using-select-columns-from-FE-to-BE-squashed' of github.com:PostHog/posthog into refactor/using-select-columns-from-FE-to-BE-squashed

Commit:
Merge remote-tracking branch 'origin/master' into refactor/using-select-columns-from-FE-to-BE-squashed

Commit:
feat: add MarketingAnalyticsTable with pinned columns to DataTable examples and stories

- Introduced MarketingAnalyticsTablePinnedColumns example in DataTable.examples.ts.
- Added corresponding story for MarketingAnalyticsTable in DataTable.stories.tsx.
- Updated DataTable.tsx to include MarketingAnalyticsTableQuery in the query options.

Commit:
refactor: remove isLastSticky from getStickyColumnInfo and related tests

- Removed isLastSticky from PinnedColumnInfo interface and its usage in getStickyColumnInfo function.
- Updated tests in columnUtils.test.tsx to reflect the removal of isLastSticky checks.
- Adjusted LemonTable and TableRow components to remove references to isLastSticky.
- Modified LemonTable.scss for padding adjustments.

Commit:
feat: update MarketingAnalyticsTable examples and stories

- Renamed MarketingAnalyticsTablePinnedColumns to MarketingAnalyticsTablePinnedColumnsAtTheBeginning.
- Added new example MarketingAnalyticsTablePinnedColumnsInTheMiddle with additional columns.
- Updated corresponding stories in DataTable.stories.tsx to reflect the changes.

lricoy reviewed (approved):
Nicely done!
